src/transformers/modeling_attn_mask_utils.py

```diff
@@ -260,13 +260,18 @@ def _ignore_causal_mask_sdpa(
 
         ignore_causal_mask = False
 
+        # Unfortunately, for query_length > 1 and key_value_length != query_length, we cannot generally ignore the attention mask, as SDPA causal mask generation
+        # may be wrong. We will set `is_causal=False` in SDPA and rely on Transformers attention_mask instead, hence not setting it to None here.
+        # Reference: https://github.com/pytorch/pytorch/issues/108108
         if attention_mask is None:
-            # TODO: When tracing with TorchDynamo with fullgraph=True, the model is recompiled depending on the input shape, thus SDPA's `is_causal` argument is rightfully updated (see https://gist.github.com/fxmarty/1313f39037fc1c112508989628c57363). However, when using `torch.export` or
-            # or `torch.onnx.dynamo_export`, we must pass an example input, and `is_causal` behavior is hard-coded. If a user exports a model with q_len > 1, the exported model will hard-code `is_causal=True` which is in general wrong (see https://github.com/pytorch/pytorch/issues/108108).
-            # Thus, we currently can NOT set `ignore_causal_mask = True` here. We would need a `torch._dynamo.is_exporting()` flag.
-            #
-            # Besides, jit.trace can not handle the `q_len > 1` condition for `is_causal` (`TypeError: scaled_dot_product_attention(): argument 'is_causal' must be bool, not Tensor`).
-            if sliding_window is None or key_value_length < sliding_window:
+            if (query_length == 1 or key_value_length == query_length) and (
+                sliding_window is None or key_value_length < sliding_window
+            ):
```

Comment:
if we are not tracing:
- if there is no padding mask, but qlen=1 or k = q or klen < sliding: we can ignore the mask because bidirectional <=> causal. 
- if padding mask, but sliding window or whatever but no pad token: we can ignore.

Reply:
> Could you pin point which part was not taken into account for speculative decoding? qlen==1 + qlen == klen?

yes

Reply:
made it cleaner